### PR TITLE
Allow disabling sriov_drivers_autoprobe

### PR DESCRIPTION
**Pull request by rjarry**
By default, VFs are automatically bound to their default kernel driver after being created.  This leads to creating a new network interface (linux netdev) for every VF.  Every netdev consumes IRQs (for RX/TX queues and other operations).  This can cause a very large number of IRQ to be created alongside which all need to be handled by a host CPU. A single VF may consume more than 10 IRQs, depending on the driver.

Intel CPU APICs can only handle a limited amount of IRQ.  Basically, there are 256 interrupt vectors in the APIC.  55 of them are reserved for general purpose interrupts (exceptions, division by zero, etc.) That leaves 201 of them for hardware/device related MSI.

Using cpu_affinity, it is possible to control which IRQs are handled by which CPUs, up to this 201 limit.  Past that limit, the kernel will fail to map the IRQ to a CPU if its APIC is already full.  This will cause the IRQs to be bound to the next available CPU.

In OpenStack NFV deployments, it is frequent to use CPU isolation to ensure that no IRQ nor CPU steal can happen on cores that are dedicated for virtual machines.  If there are too many IRQs to pack in housekeeping cores, there will be overspilling in CPUs that are expected to be isolated.  This can lead to workload interruptions and in the end cause latency and performance issues.

Additionally, even if IRQ overspilling is not an issue, all created VFs associated with a linux network interface will be handled by NetworkManager by default (unless some explicit configuration has been created to force NM to ignore them).  NM will try to configure each VF using DHCP by default.  This can cause excessive netlink churn and even potential stalling of DPDK based applications that rely on bifurcated drivers (e.g. mlx5).

In most cases, the VFs created by os-net-config are only meant for virtual machines and are no use to the host.  When the VF devices are assigned to a guest, they need to be bound to the vfio-pci driver anyway, effectively unbinding them from their default kernel driver. This in turn removes their associated linux netdev (and their related IRQs).  However, until they are bound to a guest, the unused VFs consume IRQs and can cause issues as described above.

Force sriov_drivers_autoprobe to 0 (the default value is 1) before setting sriov_numvfs.  This causes VF PCI devices to be created but prevents the kernel from binding them to their respective default driver, thus preventing the creation of netdevs and avoiding the need for CPUs to handle extraneous IRQs.

Link: https://bugzilla.redhat.com/2184735
Link: https://docs.kernel.org/PCI/pci-iov-howto.html#sr-iov-api

### CODE REVIEW

**Comment by nhicher** [comment]
Hello @rjarry,

Can you remove the depends-on for it break zuul on review.rdo. We hit github rate-limit for our github application is not present on nmstate repo.

Thanks

**Comment by rjarry** [comment]
Hi @nhicher, I have changed it to `Link:` but the code actually depends on https://github.com/nmstate/nmstate/pull/2643. Without it, some checks will fail.

**Comment by rjarry** [comment]
@karthiksundaravel could you have another look?

Thanks!

**Review by karthiksundaravel — APPROVED** [review]
Thanks

File: os_net_config/sriov_config.py [hunk 1]
```diff
@@ -827,7 +877,13 @@ def main(argv=sys.argv, main_logger=None):
     if opts.numvfs:
         if re.match(r"^\w+:\d+$", opts.numvfs):
             device_name, numvfs = opts.numvfs.split(':')
-            set_numvfs(device_name, int(numvfs))
+            pfs = common.get_sriov_map(device_name)
+            if pfs:
+                autoprobe = pfs[0]['drivers_autoprobe']
```
Comment (karthiksundaravel) [inline os_net_config/sriov_config.py:882]: please add a check if the dict has 'drivers_autoprobe' key

**Review by vcandapp — CHANGES_REQUESTED** [review]

File: os_net_config/sriov_config.py [hunk 2]
```diff
@@ -161,6 +165,50 @@ def _wait_for_vf_creation(pf_name, numvfs):
     logger.info(f"{pf_name}: Required VFs are created")
 
 
+def get_drivers_autoprobe(ifname):
```
[the remark this answers is not in the document]
Comment (rjarry) [reply os_net_config/sriov_config.py:168]: I don't know about switchdev. It probably requires the VF to be probed but I could be wrong. I have not tested it.

**Review by vcandapp — APPROVED** [review]
lgtm

File: os_net_config/sriov_config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vcandapp) [reply os_net_config/sriov_config.py:169]: As discussed, we will add some release notes like: "This flag should not be set for switchdev / vdpa / nicpartionined" devices"...

**Review by karthiksundaravel — APPROVED** [review]
lgtme. Thanks

**Pull request merged**